includes/course-video/blocks/class-sensei-course-video-blocks-youtube-extension.php

```diff
@@ -0,0 +1,147 @@
+<?php
+/**
+ * File containing the Sensei_Course_Video_Blocks_Youtube_Extension class.
+ *
+ * @package sensei-lms
+ * @since 4.0.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Extends standard Embed block with YouTube specific functionality for video course progression
+ *
+ * @since 4.0.0
+ */
+class Sensei_Course_Video_Blocks_Youtube_Extension {
+	/**
+	 * Instance of class.
+	 *
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * Manages Video course-related settings.
+	 *
+	 * @var Sensei_Course_Video_Settings
+	 */
+	private $settings;
+
+	/**
+	 * Returns an instance of the class.
+	 *
+	 * @param Sensei_Course_Video_Settings $settings
+	 *
+	 * @return Sensei_Course_Video_Blocks_Youtube_Extension
+	 */
+	public static function instance( Sensei_Course_Video_Settings $settings ) {
+		if ( self::$instance ) {
+			return self::$instance;
+		}
+
+		self::$instance = new self( $settings );
+		return self::$instance;
+	}
+
+	/**
+	 * Initialize the class and hooks.
+	 *
+	 * @param Sensei_Course_Video_Settings $settings
+	 */
+	public static function init( Sensei_Course_Video_Settings $settings ) {
+		self::instance( $settings )->init_hooks();
+	}
+
+	/**
+	 * Sensei_Youtube_Extension constructor.
+	 *
+	 * @param Sensei_Course_Video_Settings $settings
+	 */
+	private function __construct( Sensei_Course_Video_Settings $settings ) {
+		$this->settings = $settings;
+	}
+
+	/**
+	 * Initialize hooks.
+	 */
+	public function init_hooks() {
+		add_filter( 'embed_oembed_html', [ $this, 'wrap_youtube' ], 10, 4 );
+	}
+
+	/**
+	 * Wrap YouTube video in a container.
+	 *
+	 * @param string $html
+	 * @param string $url
+	 * @param array  $args
+	 * @param int    $post_id
+	 *
+	 * @return string
+	 */
+	public function wrap_youtube( $html, $url, $args, $post_id ) {
+		if ( ! $this->is_youtube_url( $url ) ) {
+			return $html;
+		}
+
+		$this->enqueue_scripts();
```

Comment:
Depending on the solution you find for the cache, it might be needed to consider enqueueing it in some other way instead of depending on this filter.
A reference here with the block check: https://wordpress.stackexchange.com/questions/328536/load-css-javascript-in-frontend-conditionally-if-block-is-used